src/rgw/rgw_pubsub.cc

```diff
@@ -502,46 +502,62 @@ RGWPubSub::RGWPubSub(rgw::sal::Driver* _driver, const std::string& _tenant)
   : driver(_driver), tenant(_tenant)
 {}
 
-int RGWPubSub::read_topics(const DoutPrefixProvider *dpp, rgw_pubsub_topics& result,
-    RGWObjVersionTracker *objv_tracker, optional_yield y) const
+int RGWPubSub::get_topics(const DoutPrefixProvider* dpp,
+                          const std::string& start_marker, int max_items,
+                          rgw_pubsub_topics& result, std::string& next_marker,
+                          optional_yield y) const
 {
-  if (driver->get_zone()->get_zonegroup().supports_feature(
+  if (!driver->get_zone()->get_zonegroup().supports_feature(
           rgw::zone_features::notification_v2)) {
-    void* handle = NULL;
-    auto ret =
-        driver->meta_list_keys_init(dpp, "topic", std::string(), &handle);
-    if (ret < 0) {
-      return ret;
-    }
-    bool truncated;
-    int max = 1000;
-    do {
-      std::list<std::string> topics;
-      ret = driver->meta_list_keys_next(dpp, handle, max, topics, &truncated);
-      if (ret < 0) {
-            ldpp_dout(dpp, 1)
-                << "ERROR: lists_keys_next(): " << cpp_strerror(-ret) << dendl;
-            break;
-      }
-      for (auto& topic_entry : topics) {
-        std::string topic_name;
-        std::string topic_tenant;
-        parse_topic_entry(topic_entry, &topic_tenant, &topic_name);
-        if (tenant != topic_tenant) {
-          continue;
-        }
-        rgw_pubsub_topic topic;
-        const auto op_ret = get_topic(dpp, topic_name, topic, y, nullptr);
-        if (op_ret < 0) {
-          ret = op_ret;
-          continue;
-        }
-        result.topics[topic_name] = std::move(topic);
-      }
-    } while (truncated);
-    driver->meta_list_keys_complete(handle);
+    // v1 returns all topics, ignoring marker/max_items
+    return read_topics_v1(dpp, result, nullptr, y);
+  }
+
+  // TODO: prefix filter on 'tenant:'
+  void* handle = NULL;
+  int ret = driver->meta_list_keys_init(dpp, "topic", start_marker, &handle);
+  if (ret < 0) {
+    return ret;
+  }
+  auto g = make_scope_guard(
+      [this, handle] { driver->meta_list_keys_complete(handle); });
+
+  if (max_items > 1000) {
+    max_items = 1000;
```

Comment:
if `max_items` is more than 1000, we still explicitly return only 1000 entries, or probably we should `do-while` to list more than 1000 as caller is requesting for more than 1000 items ?

Reply:
the do-while for radosgw-admin is in rgw_admin.cc under OPT::PUBSUB_TOPIC_LIST

Reply:
> the do-while for radosgw-admin is in rgw_admin.cc under OPT::PUBSUB_TOPIC_LIST

yeah it does, however it does not obey the max_entries specified, what that means is if the radosgw-admin `max_entries` is any value > 1000 but less the total_entries_count  it will get back all the entries, the code does not stop at `max_entries` value
not a big deal though

Reply:
gotcha, thanks. i added special handling for `max_entries_specified` so it stops looping there, but by default (no --max-entries) it'll list everything. i also hooked up `--marker` to resume the listing. testing that, i found/fixed a bug with next_marker in `RGWPubSub::get_topics()`